### DISCUSSION TITLE
Stuck on setting up with Python backend.

### DISCUSSION THREAD

**Discussion by JamesWingert** [nextauthjs/next-auth#4341] [Help]
Hello, 
I am currently learning auth and nextjs in general. I have set up a Python(FastAPI) backend. The OpenAPI docs and everything are working correctly from the Python perspective. Here is a preview of the successful login. 
![Screen Shot 2022-04-05 at 12 13 52 PM](https://user-images.githubusercontent.com/44770511/161832287-1d2ce28f-ce06-4e4f-98fd-a735e0444383.png)

In addition, when I run the command in my prompt, as seen here on step 2 for login:
[https://fastapi-users.github.io/fastapi-users/usage/flow/#conclusion](url)

It also lets me login successfully.

I am unable to use the SignIn feature using the Credential Provider. I have been beating my head against a wall for the last 3 days trying to figure this out. I'm sure it's just something that I am not understanding correctly but I would really appreciate any help as I have been stuck for so long on this and can only find a handful of (seemingly outdated)  Django examples. I have tried everything that I have come across and am super stuck. 

Current attempt error:
`Uncaught (in promise) SyntaxError: Unexpected token < in JSON at position VM2266:1 0`

When I go to VM2266:1 it is an HTML document that starts with 
`<!DOCTYPE html>`
Which highlights the first < as the error.

When I sign in from .../schools3 I get:

`Error: Callback for provider type credentials not supported`
and
`{"credentials":{"id":"credentials","name":"Credentials","type":"credentials","signinUrl":"http://localhost:3000/api/auth/signin/credentials","callbackUrl":"http://localhost:3000/api/auth/callback/credentials"}}`

When I sign in from ../api/auth/signin
`http://localhost:3000/api/auth/signin?csrf=true`
It gets stuck there and gives me a csrf token.

I have a feeling it has something to do with the fact that the auth/jwt/login requires form rather than JSON but I cannot figure out how to implement that correctly.
```
///api/auth/[...nextauth].ts

/* eslint-disable @typescript-eslint/ban-ts-comment */
import axios from 'axios';
import FormData from 'form-data';
import NextAuth from 'next-auth';
import CredentialsProvider from 'next-auth/providers/credentials';
export const config = {
  api: {
    bodyParser: false,
  },
};
export default NextAuth({
  providers: [
    CredentialsProvider({

      credentials: {
        username: {
          label: 'username',
          type: 'text',
        },
        password: { label: 'Password', type: 'password' },
      },
      //@ts-ignore
      async authorize(credentials: any, req) {

        const formData = new FormData();
        //@ts-ignore
        formData.append('username', `${credentials.username}`);
        //@ts-ignore
        formData.append('password', `${credentials.password}`);
        const res = axios
          .post('http://localhost:8000/auth/jwt/login', formData, {
            headers: {
              'Content-Type': 'multipart/form-data',
            },
          })
          .then(response => response)
          .catch(error => error)
        //@ts-ignore
        const user = await res.data.json();
        // If no error and we have user data, return it
        //@ts-ignore
        if (res.data.data.access_token && user) {
          // debug
          return {
            user,
          };
        }

        // Return null if user data could not be retrieved
        return null;
      },
    }),
    // ...add more providers here
  ],
  session: {
    strategy: 'jwt',
  },

  callbacks: {
    async jwt({ token, user }) {
      if (user?.token) {
        // @ts-ignore
        token = { access_token: user.access_token };
      }
      return token;
    },
    async session({ session, token, user }) {
      session.access_token = token.access_token;
      return session;
    },
  },

  debug: true,
});
```
```
///pages/schools3.tsx

import type { NextPage } from 'next';
import { signIn, signOut, useSession } from 'next-auth/react';

import { DynamicHeader } from '@/components/Header';

const SignInOutButton = ({ isSignedIn }: { isSignedIn: boolean }) => {
  if (isSignedIn) {
    return <button onClick={() => signOut()}>Sign Out</button>;
  }
  return (
    <button
      onClick={() => signIn('credentials')}
      // 'credentials', { callbackUrl: 'http://localhost:3000/schools3' }
    >
      Sign in
    </button>
  );
};

const Schools3: NextPage = () => {
  const { data: session, status } = useSession();
  const authData = {
    session,
    status,
  };

  return (
    <>
      <DynamicHeader authData={authData} />

      <div
        style={{
          display: 'flex',
          flexDirection: 'column',
          maxWidth: '100px',
        }}
      >
        <SignInOutButton isSignedIn={!!session?.user} />
      </div>
    </>
  );
};

export default Schools3;
```

Thank you again for any help you can provide I really really appreciate it.

**Comment by movaldivia** [comment]
@JamesWingert did you found the problem?

**Comment by Amity808** [comment]
@JamesWingert were you able to get the solution, I got stock as well

**Comment by yutian-wu** [comment]
I ran into a similar error before, and I believe it's because the ``callbackUrl`` uses `http` as its protocol.

My guess is that using a plain text protocol for authentication callback is unsafe, since the credential / token will be in the callback request. That's why `next-auth` doesn't support it and throws that error, though I think it's okay for `dev` mode.

One workaround is to set up an Nginx server that serves as a TLS-termination proxy for the web server and the back-end server, and then use `https://localhost/[callback-path]` as `callbackUrl`.

- Note that there's no port in the URL, since we want the requests to hit Nginx (HTTPS) server first.

Since you have a separate back-end server running, you may configure Nginx to redirect requests based on their paths, for example:
```
location / {
  proxy_pass http://localhost:3000;
}
location /auth/jwt/login {
  proxy_pass http://localhost:8000;
}
```
